### ISSUE TITLE
Entities not showing 

### ISSUE THREAD

**Issue by RhinoRich** [gazoodle/gecko-home-assistant#9]
## Version of the custom_component
Custom component V0.0.5
![image](https://user-images.githubusercontent.com/76866461/107878362-04541300-6eca-11eb-94ba-d438a59f1ef6.png)

## Configuration

```

# Configure a default setup of Home Assistant (frontend, api, etc)
default_config:

# Text to speech
tts:
  - platform: google_translate

group: !include groups.yaml
automation: !include automations.yaml
script: !include scripts.yaml
scene: !include scenes.yaml

climate:
  - platform: generic_thermostat
    name: heat_control
    heater: switch.lidl_socket_1
    target_sensor: sensor.climate_temperature
    min_cycle_duration:
      minutes: 5


```

## Describe the bug
Entities not showing in integration

## Debug log

```
2021-02-14 13:53:25 INFO (MainThread) [custom_components.gecko]
-------------------------------------------------------------------
Gecko
Version: 0.0.1
This is a custom integration!
If you have any issues with this you need to open an issue here:
https://github.com/gazoodle/gecko-home-assistant/issues
-------------------------------------------------------------------
2021-02-14 13:53:25 INFO (MainThread) [custom_components.gecko] Setup entry for SPA04:91:62:d3:2b:26
2021-02-14 13:53:26 INFO (MainThread) [custom_components.gecko] Locator GeckoLocator(age=1.1189897650001512,has_had_enough_time=False,spas=[Ducky Tub(SPA04:91:62:d3:2b:26)]) ready
2021-02-14 13:53:26 INFO (MainThread) [custom_components.gecko] Waiting for facade to be ready
2021-02-14 13:53:34 ERROR (Thread-6) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.8/threading.py", line 870, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/udp_socket.py", line 364, in _thread_func
self._loop_func()
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 260, in _loop_func
self._final_connect()
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 264, in _final_connect
self.struct.build_accessors([self.config_xml, self.log_xml])
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spastruct.py", line 72, in build_accessors
self.accessors = {
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spastruct.py", line 75, in <dictcomp>
for element in xml.findall(
AttributeError: 'NoneType' object has no attribute 'findall'
2021-02-14 13:53:56 ERROR (Thread-5) [root] Uncaught thread exception
Traceback (most recent call last):
File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
self.run()
File "/usr/local/lib/python3.8/threading.py", line 870, in run
self._target(*self._args, **self._kwargs)
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 282, in _ping_thread_func
self.refresh()
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 311, in refresh
if not self.is_connected:
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 306, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
2021-02-14 13:53:56 ERROR (MainThread) [custom_components.gecko] Exception during entry setup
Traceback (most recent call last):
File "/config/custom_components/gecko/__init__.py", line 47, in async_setup_entry
while not facade.is_connected:
File "/usr/local/lib/python3.8/site-packages/geckolib/automation/facade.py", line 68, in is_connected
if not self._spa.is_connected:
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 306, in is_connected
raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...
```
![image](https://user-images.githubusercontent.com/76866461/107878225-0ec1dd00-6ec9-11eb-8c37-45bf20bdf258.png)

![image](https://user-images.githubusercontent.com/76866461/107878313-898af800-6ec9-11eb-9a0f-4e3d37cb926e.png)

Tried adding and removing Options. Deleting and reinstalling. Powercycling the RPi. Same with Gecko controller and Gecko Wifi-RF bridge. Tried taking the App off the network and repeating the foregoing in case of conflict. 

If I change reload the iOS app and change the spa name, then reload HA, the new name appears on the integration.

Any ideas very welcome.

**Comment by gazoodle** [comment]
Sorry for the tardyt reply, I'm not sure what's going on here, and I'm on another project at the moment. I've got some time coming up soon that I can take a look at this, probably need some more diags added to the component to help with this.

**Comment by RhinoRich** [comment]
Thanks Gazoodle. I can try out whatever you need. Would be great to get the spa integrated into my automations.

**Comment by RhinoRich** [comment]
Logger: custom_components.gecko
Source: custom_components/gecko/__init__.py:45
Integration: Gecko (documentation, issues)
First occurred: 7:38:33 PM (1 occurrences)
Last logged: 7:38:33 PM

Exception during entry setup
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 1693, in feed
    self.parser.Parse(data, 0)
xml.parsers.expat.ExpatError: junk after document element: line 20668, column 13

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/gecko/__init__.py", line 45, in async_setup_entry
    facade = locator.get_spa_from_identifier(spa_identifier).get_facade(False)
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 46, in get_facade
    facade = GeckoFacade(GeckoSpa(self).start_connect())
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 73, in __init__
    GeckoSpaPack.__init__(self)
  File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spapack.py", line 16, in __init__
    self._spa_pack_struct = ET.parse(GeckoConstants.SPA_PACK_STRUCT_FILE)
  File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 1202, in parse
    tree.parse(source, parser)
  File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 601, in parse
    parser.feed(data)
  File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 1695, in feed
    self._raiseerror(v)
  File "/usr/local/lib/python3.8/xml/etree/ElementTree.py", line 1602, in _raiseerror
    raise err
  File "<string>", line None
xml.etree.ElementTree.ParseError: junk after document element: line 20668, column 13

**Comment by awakefie** [comment]
Having the same issue.  Happy to help you debug the issue any way I can.

**Comment by awakefie** [comment]
> Having the same issue. Happy to help you debug the issue any way I can.

Also, happy to float some cash your way and a bounty!  ($100.00 US)

Specifically on this one and the IP address option.  Either way thank you for the work you have already done!

**Comment by RhinoRich** [comment]
Also happy to contribute. Being able to set the temperature setpoint based on use and electricity tariff windows would be awesome, and synching the lights to the rest of my automations would be cool, also.

Basically, having a low-cost energy window with my electricity supplier is pointless if I can't control when it's heating.

**Comment by simonhoole** [comment]
HI, I am new to HA and I am having the problem with not showing the entities. I dont understand what you have done to resolve the matter. Any chance of helping and doing a dummies / easier explanation. Sorry and thank you

**Comment by RhinoRich** [comment]
> HI, I am new to HA and I am having the problem with not showing the entities. I dont understand what you have done to resolve the matter. Any chance of helping and doing a dummies / easier explanation. Sorry and thank you

I followed the instructions to modify the spapackstruct.xml file and I think it was successful, but I get the errors shown above. I've gone way past my own level of competence to do that so I'm now at a loss. We can only hope that @gazoodle gets some time soon to help us.

**Comment by BachaSport** [comment]
Happy to contribute as well $$

I am stuck on not discovering entities. @gazoodle please give us a hand. Thank you.

**Comment by BachaSport** [comment]
this is wat i get in logs:

Logger: root
Source: /usr/src/homeassistant/homeassistant/bootstrap.py:310
First occurred: April 23, 2021, 6:36:42 PM (1 occurrences)
Last logged: April 23, 2021, 6:36:42 PM

Uncaught thread exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.8/site-packages/geckolib/driver/udp_socket.py", line 364, in _thread_func
    self._loop_func()
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 260, in _loop_func
    self._final_connect()
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 264, in _final_connect
    self.struct.build_accessors([self.config_xml, self.log_xml])
  File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spastruct.py", line 72, in build_accessors
    self.accessors = {
  File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spastruct.py", line 75, in <dictcomp>
    for element in xml.findall(
AttributeError: 'NoneType' object has no attribute 'findall'



Logger: homeassistant.bootstrap
Source: bootstrap.py:392
First occurred: April 23, 2021, 6:35:54 PM (2 occurrences)
Last logged: April 23, 2021, 6:36:54 PM

Waiting on integrations to complete setup: automation
Waiting on integrations to complete setup: gecko, automation

**Comment by BachaSport** [comment]
@gazoodle do you think you could spare couple of mintures to have a look at this please? I would really appreciate and bounty this  issue. Looking forward to your reply.

**Comment by BachaSport** [comment]
Also this one:

Logger: root
Source: /usr/src/homeassistant/homeassistant/bootstrap.py:310
First occurred: April 23, 2021, 6:37:07 PM (1 occurrences)
Last logged: April 23, 2021, 6:37:07 PM

Uncaught thread exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 282, in _ping_thread_func
    self.refresh()
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 311, in refresh
    if not self.is_connected:
  File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 306, in is_connected
    raise RuntimeError("Spa took too long to connect ...")
RuntimeError: Spa took too long to connect ...

**Comment by orangelizard** [comment]
Following. Same issue here, spa is discovered correctly but no entitys created. No other devices (phones tablets etc) connected, spa hub restarted. This is the error in the log.

Logger: custom_components.gecko
Source: util/async_.py:139
Integration: Gecko (documentation, issues)
First occurred: 19:11:07 (1 occurrences)
Last logged: 19:11:07

Exception during entry setup
Traceback (most recent call last):
File "/config/custom_components/gecko/init.py", line 45, in async_setup_entry
facade = locator.get_spa_from_identifier(spa_identifier).get_facade(False)
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 46, in get_facade
facade = GeckoFacade(GeckoSpa(self).start_connect())
File "/usr/local/lib/python3.8/site-packages/geckolib/spa.py", line 73, in init
GeckoSpaPack.init(self)
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spapack.py", line 15, in init
GeckoSpaPack.download_if_needed()
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spapack.py", line 32, in download_if_needed
GeckoSpaPack.download()
File "/usr/local/lib/python3.8/site-packages/geckolib/driver/spapack.py", line 39, in download
response = http.request(
File "/usr/local/lib/python3.8/site-packages/urllib3/request.py", line 74, in request
return self.request_encode_url(
File "/usr/local/lib/python3.8/site-packages/urllib3/request.py", line 96, in request_encode_url
return self.urlopen(method, url, **extra_kw)
File "/usr/local/lib/python3.8/site-packages/urllib3/poolmanager.py", line 375, in urlopen
response = conn.urlopen(method, u.request_uri, **kw)
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 699, in urlopen
httplib_response = self._make_request(
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 394, in _make_request
conn.request(method, url, **httplib_request_kw)
File "/usr/local/lib/python3.8/site-packages/urllib3/connection.py", line 234, in request
super(HTTPConnection, self).request(method, url, body=body, headers=headers)
File "/usr/local/lib/python3.8/http/client.py", line 1255, in request
self._send_request(method, url, body, headers, encode_chunked)
File "/usr/local/lib/python3.8/http/client.py", line 1266, in send_request
self.putrequest(method, url, **skips)
File "/usr/local/lib/python3.8/site-packages/urllib3/connection.py", line 214, in putrequest
return HTTPConnection.putrequest(self, method, url, *args, **kwargs)
File "/usr/src/homeassistant/homeassistant/util/async.py", line 150, in protected_loop_func
check_loop()
File "/usr/src/homeassistant/homeassistant/util/async.py", line 139, in check_loop
raise RuntimeError(
RuntimeError: I/O must be done in the executor; Use await hass.async_add_executor_job() at custom_components/gecko/init.py, line 45: facade = locator.get_spa_from_identifier(spa_identifier).get_facade(False)

I'm quite a noob to home assistant, any ideas?

**Comment by BachaSport** [comment]
nope, need to wait for @gazoodle to help us out here.

**Comment by orangelizard** [comment]
👍🏻All good things come to those who wait 🙂

**Comment by RhinoRich** [comment]
From what I understand of gazoodle's posts in the last few months, interpreting the in.YT and latest SpaPackStruct isn't a trivial job, particularly as he doesn't have that model of Spa Pack himself. He (and we) really need Gecko to be a bit more helpful.

**Comment by maegibbons** [comment]
@awakefie @BachaSport @orangelizard 

Did you get yourselves sorted out?

I have helped out @RhinoRich with his setup now and thats working with InYT.

Would need to get ssh access to your HA box but I can talk you through that.

Krs

Mark

**Comment by orangelizard** [comment]
Hi Mark, I've never managed to sort this out. Is there a guide posted somewhere? 

Kind regards, Jason

**Comment by RhinoRich** [comment]
I gave Mark access to HA remotely and he implemented the bug fixes and modified SpaPackStruct that way. Works like a charm!

My only niggle now is lovelace overriding the 'toggle' tap-action for the jet icons with 'more-info' which is really annoying.
![image](https://user-images.githubusercontent.com/76866461/121645366-056a0b00-ca8c-11eb-983c-36528ddec9f7.png)

**Comment by maegibbons** [comment]
@orangelizard

As @RhinoRich said I need to access  your  HA  system to check Spa revisions 

Hit me up on https://m.facebook.com/mark.ae.gibbons initially to exchange email addresses and instructions.

Krs

Mark

**Comment by maegibbons** [comment]
Or look at the web link domain in my profile and email me at mgibbons@<weblink.domain>

**Comment by BachaSport** [comment]
> @orangelizard
> 
> As @RhinoRich said I need to access your HA system to check Spa revisions
> 
> Hit me up on https://m.facebook.com/mark.ae.gibbons initially to exchange email addresses and instructions.
> 
> Krs
> 
> Mark

Hello @maegibbons, i just send you my email on FB. I would be grateful of you can give me some instructions or look into my HA to get this going :) Thanks in advance!

**Comment by gazoodle** [comment]
Sorry I’ve not responded on this thread, reading through all your posts, I think they are mostly incarnations of the wrong SpaStructPackxml issue just manifesting itself in different ways (but also mixed with some static IP requests) If this is the case, then v0.0.10 ought to fix them, so please can you take a look when you are able and if it does fix it, then please let me know so I can close the issue.

oh, and thanks for the offer of a bounty, it’s much appreciate but not needed, this is my way of giving back to the HA community for all … sorry it’s taken me so long to get back to this.

**Comment by orangelizard** [comment]
Thanks gazoodle, I have just tried installing and all went fine. So working for me. Thank you for your time and contribution. Kind regards, orangelizard.

**Comment by gazoodle** [comment]
Confirmed it's fixed, so closing.
